lib/roo_on_rails/rack/populate_env_from_jwt.rb

```diff
@@ -0,0 +1,98 @@
+require 'json/jwt'
+require 'faraday'
+require 'faraday_middleware'
+
+module RooOnRails
+  module Rack
+    class PopulateEnvFromJWT
+      UnnacceptableKeyError = Class.new(RuntimeError)
+      VALID_JWK_URL_PREFIXES = [
+        'https://www.deliveroo.com/identity-keys/',
+        'https://identity.deliveroo.com/identity-keys/'
+      ].freeze
+
+      def initialize(app, logger:)
+        @app = app
+        @keys = {}
+        @logger = logger
+
+        if development?
+          @logger.warn "JWTs will not have their signatures checked, you're in development mode."
+        end
+      end
+
+      def call(env)
+        env['roo.identity'] = decode_authorization_header(env['HTTP_AUTHORIZATION'])
+        @app.call(env)
+
+      # Other exceptions will bubble up, allowing the higher middleware to return a 500, which is
+      # intentional.
+      rescue UnnacceptableKeyError, JSON::JWT::Exception => e
+        # Identifying user is clearly attempting to hack or has been given a totally incorrect
+        # token, log this and flag as Forbidden, without executing the rest of the middleware stack.
+        ::NewRelic::Agent.notice_error(e) if defined?(NewRelic)
+        [403, {}, []]
```

Comment:
This should really be a `401`:

* `401 Unauthorized` - I either don't know who you are or don't _believe_ you are who you say you are.
* `403 Forbidden` - I know and believe who you are, but you cannot do what you're trying to do.

Reply:
Also, this may have further-reaching concerns than you expect. If the page the user wants to visit doesn't require authentication then this probably shouldn't fail just because the credentials they're presenting are invalid. It could reasonably just ignore them.

Reply:
Ahh; for some reason I thought 403 was "no matter what you do, you can't do this" (because you're trying to do nefarious things with your dodgy token). I'll make the change :)

Reply:
To your second comment: I considered just passing "no claims" through to the rest of the middleware stack; but the only cases where this block should be executed is if the token _could never_ have been correct (ie. maliciously crafted, or something has been incorrectly configured) — which is why it pings NewRelic and gives up.

Reply:
The key difference is that with a 401 reauthenticating might help (which here it might; maybe somehow your token got corrupted and you could reauthenticate to get a new one). With a 403 reauthenticating will never help as the system already knows who you are and you're not allowed.